src/Data/HashMap/Linear.hs

```diff
@@ -120,141 +149,255 @@ data ProbeResult k v where
-          shiftSegmentBackward cap arr'' ((ix + 1) `mod` cap) & \arr''' ->
-            HashMap (Count (c - 1)) arr'''
-
--- | Shift all cells with PSLs > 0 in a continuous segment
--- following the deleted cell, backwards by one and decrement
--- their PSLs.
-shiftSegmentBackward :: Keyed k =>
-  Int -> RobinArr k v #-> Int -> RobinArr k v
-shiftSegmentBackward s arr ix = Array.read arr ix & \case
-  (arr', Ur Nothing) -> arr'
-  (arr', Ur (Just (RobinVal 0 _ _))) -> arr'
-  (arr', Ur (Just val)) ->
-    Array.write arr' ix Nothing & \arr'' ->
-      shiftSegmentBackward
-        s
-        (Array.write arr'' ((ix-1) `mod` s) (Just $ decRobinValPSL val))
-        ((ix+1) `mod` s)
+delete hm k = alter hm k (\_ -> Nothing)
 
--- | 'insert' (in the provided order) a list of key-value pairs
--- to a given hashmap.
+-- | 'insert' (in the provided order) the given key-value pairs to
+-- the hashmap.
 insertAll :: Keyed k => [(k, v)] -> HashMap k v #-> HashMap k v
 insertAll [] hmap = hmap
 insertAll ((k, v) : xs) hmap = insertAll xs (insert hmap k v)
+-- TODO: Do a resize first on the length of the input.
 
--- | Given a key, psl of the probe so far, current unread index, and
--- a full hashmap, return a probe result: the place the key already
--- exists, a place to swap from, or an unfilled cell to write over.
-probeFrom :: Keyed k =>
-  (k, PSL) -> Int -> HashMap k v #-> (HashMap k v, ProbeResult k v)
-probeFrom (k, p) ix (HashMap ct arr) = Array.read arr ix & \case
-  (arr', Ur Nothing) ->
-    (HashMap ct arr', IndexToInsert p ix)
-  (arr', Ur (Just robinVal'@(RobinVal psl k' v'))) ->
-    case k == k' of
-      -- Note: in the True case, we must have p == psl
-      True -> (HashMap ct arr', IndexToUpdate v' psl ix)
-      False -> case psl < p of
-        True -> (HashMap ct arr', IndexToSwap robinVal' p ix)
-        False ->
-          capacity (HashMap ct arr') & \(HashMap ct' arr'', Ur cap) ->
-            probeFrom (k, p+1) ((ix+1)`mod` cap) (HashMap ct' arr'')
+-- | A version of 'fmap' which can throw out the elements.
+--
+-- Complexity: O(capacity hm)
+mapMaybe :: Keyed k => (v -> Maybe v') -> HashMap k v #-> HashMap k v'
+mapMaybe f = mapMaybeWithKey (\_k v -> f v)
+
+-- | Same as 'mapMaybe', but also has access to the keys.
+mapMaybeWithKey :: Keyed k => (k -> v -> Maybe v') -> HashMap k v #-> HashMap k v'
+mapMaybeWithKey (f :: k -> v -> Maybe v') (HashMap _ arr') =
+  Array.size arr' & \case
+    (arr'', Ur 0) ->
+      recurOrReturn True 0 0 0 0 arr'' & \(arr''', Ur _) ->
+        HashMap 0 arr'''
+    (arr'', Ur sz) ->
+      numSpillovers arr'' & \(arr''', Ur spillovers) ->
+        go
+          spillovers
+          ((sz - 1 + spillovers) `mod` sz)
+          sz
+          0
+          arr'''
+          & \(arr'''', Ur b) -> HashMap b arr''''
+ where
+  go :: Int -- ^ Current index
+     -> Int -- ^ Last index to check
+     -> Int -- ^ Size of the array
+     -> Int -- ^ Number of resulting elements
```

Comment:
This is an accumulated count of pairs that are present after the map. Just change the comment please.